nvfuser torchbench patch

jjsjann123 opened the pull request:
1. Patching nvfuser_execute to take aten nvprim fallback when no cuda tensors are provided as inputs
2. Extending support of nvfuser python API on cpu scalar tensor.

jjsjann123 committed:
nvprim lowering context update to avoid cpu tensor; nvfuser executor to fallback to aten for trivial fusion

jjsjann123 committed:
Merge remote-tracking branch 'upstream/viable/strict' into HEAD

jjsjann123 committed:
fixing tests

IvanYashchuk reviewed (commented):
Could you please add a test with the actual execution of a graph with CPU scalar tesnor inputs?

jjsjann123 commented:
> Could you please add a test with the actual execution of a graph with CPU scalar tesnor inputs?

Just to confirm this one, cpu scalar input is not yet supported with our python frontend. So a test with that will trigger an error and we'll need to have the test skipped. I'm thinking about adding that in the follow up PR where we actually add the support.

jjsjann123 committed:
fixing lintrunner; addressing review comments

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 committed:
enable cpu scalar for nvprims context

jjsjann123 committed:
cpu scalar tensor added

jjsjann123 committed:
added default arg

jjsjann123 committed:
cpu scalar test added

jjsjann123 committed:
dtype added

jjsjann123 committed:
missing import

jjsjann123 committed:
fixing test

jjsjann123 commented:
Added cpu scalar in the python API. cc'ing @kevinstephano

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into wip1

jjsjann123 committed:
lintrunner

jjsjann123 committed:
lintrunner

jjsjann123 commented:
Looks like CI is green now. But I'd like to get @kevinstephano 's stamp on python API changes before merging~

IvanYashchuk commented on torch/_prims/context.py:
```diff
@@ -207,10 +208,19 @@ def _is_func_unsupported_nvfuser(torch_function_mode, func, args, kwargs):
+    flat_args, _ = tree_flatten(args)
+    # TODO: allow cpu scalar once we plumb the support via python API
+    # check for unsupport cpu tensor
+    if any(isinstance(arg, torch.Tensor) and arg.is_cpu and arg.ndim != 0 for arg in flat_args):  # type: ignore[attr-defined]
```
So this line marks the node as unsupported if any of the arguments is a non-scalar CPU tensor? How can mixing devices happen without an error?
For example, addition is used in the tests, which just errors out as expected:
```py
import torch
from torch._prims.context import TorchRefsMode
a = torch.randn(3, 3, device="cuda")
b = torch.randn(3, 3, device="cpu")
# raises RuntimeError: Tensor on device cpu is not on the expected device cuda:0!
with TorchRefsMode():
    out = a + b

b = torch.tensor(0.5)
# works
with TorchRefsMode():
    out = a + b
```

IvanYashchuk replied:
Why do we need to resort to aten op when even there mixing cuda and cpu tensors should error out:
```py
In [1]: import torch

In [2]: a = torch.randn(3, 3, device='cuda')

In [3]: a + a.cpu()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 a + a.cpu()

RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

jjsjann123 replied:
The magic here is `arg.ndim != 0`.

mix device only really works with cpu scalar, in which case it's usually handled by cuda kernels that takes cpu scalar as a regular scalar.

```
In [1]: import torch

In [2]: a = torch.randn(3, 3, device="cuda")

In [3]: y = torch.tensor(0.5)

In [4]: y
Out[4]: tensor(0.5000)

In [5]: y.device
Out[5]: device(type='cpu')

In [6]: print(a + y)
tensor([[ 0.0930,  1.7815,  0.8127],
        [ 0.4306, -0.0681,  0.4698],
        [ 1.7451, -1.8241,  1.1352]], device='cuda:0')

```

IvanYashchuk replied:
Yes I see `arg.ndim == 0` is allowed.

I mean why should we have any special login for the case of `arg.ndim !=0` if it errors out also in eager and primtorch?

IvanYashchuk commented on torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h:
```diff
@@ -256,6 +258,12 @@ struct InputTensorRecord : RecordFunctor {
+    if (symbolic_sizes.empty() && is_cpu) {
+      tv->setCpuScalar(true);
+    } else {
+      TORCH_INTERNAL_ASSERT(!is_cpu, "cpu non-scalar tensor is not supported");
```
This should be `TORCH_CHECK` instead of `TORCH_INTERNAL_ASSERT` because the assert is something that's unexpected and the error message suggest to the user to file a bug report.

When we have the following graph:
```
# gm.graph.print_tabular()
opcode         name    target               args        kwargs
-------------  ------  -------------------  ----------  --------
placeholder    x_1     x_1                  ()          {}
placeholder    y_1     y_1                  ()          {}
call_function  add     nvprims.add.default  (x_1, y_1)  {}
output         output  output               (add,)      {}
```
We'd get the following error message:
```py
# nvfuser_execute(gm, a, a.cpu())
RuntimeError: !is_cpu INTERNAL ASSERT FAILED at "/home/iyashchuk/dev/pytorch/master/torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h":264, please report a bug to PyTorch. cpu non-scalar tensor is not supported
```

IvanYashchuk commented on torch/_prims/nvfuser_executor.py:
```diff
@@ -41,7 +42,10 @@ def to_nvfuser_template_args(args):
+                arg.size(),
+                arg.stride(),
+                getnvFuserDtype(arg.dtype),
+                arg.device.type == "cpu",
```
```suggestion
                arg.is_cpu,
```

jjsjann123 replied:
errr.. that `is_cpu` was what I had in the first place and it triggers a linter error. https://github.com/pytorch/pytorch/runs/8161952307?check_suite_focus=true

IvanYashchuk replied:
Skip this line for linter? 😄

IvanYashchuk commented on torch/_prims/context.py:
```diff
@@ -207,10 +208,19 @@ def _is_func_unsupported_nvfuser(torch_function_mode, func, args, kwargs):
+        return True
+
+    flat_args, _ = tree_flatten(args)
+    # TODO: allow cpu scalar once we plumb the support via python API
```
Done? `nvprims.add(cuda_tensor, cpu_scalar)` works now.

IvanYashchuk reviewed (commented):
Thank you for fixing CPU scalars, that's great!

If we now revert changes to `torch/_prims/context.py`, `test_cpu_scalar` would still pass.

jjsjann123 commented:
> If we now revert changes to torch/_prims/context.py, test_cpu_scalar would still pass.

Yep, that's how the tests are intended to behave. cpu scalar support is plumbed through python API.

Changes to `context.py` is only trying to prevent accidental fusion on cpu devices, reverting changes there should break the other test I updated.

jjsjann123 committed:
Merge commit '9a6aa9053f79127721875e371addd9c3baeaaac0' into wip1

jjsjann123 committed:
review comments

IvanYashchuk commented:
I think changes to `test_nvfuser_capability_context` should also be reverted in addition to `torch/_prims/context.py`.

Can you please provide a concrete example of what situation changes to `context.py` prevent? Because I don't see what you would like to achieve.

jjsjann123 commented:
> I think changes to `test_nvfuser_capability_context` should also be reverted in addition to `torch/_prims/context.py`.
> 
> Can you please provide a concrete example of what situation changes to `context.py` prevent? Because I don't see what you would like to achieve.

Those changes are to prevent passing cpu fusion to nvprim, which can only handle cuda kernels vvv
> 1. Patching TorchRefsNvfuserCapabilityMode to avoid lowering node to nvprim which takes non-cuda tensor as inputs

jjsjann123 committed:
Merge commit 'd2b8b8f29121ed23e2b39446d09bba3a7eb96684' into nvprim_context

jjsjann123 committed:
moving conditional logic to nvfuser_execute

jjsjann123 committed:
updating tests; add warning

jjsjann123 committed:
fixing tests with assert on warning

jjsjann123 committed:
fixing tests

jjsjann123 committed:
fixing fallback condition on cpu scalar

jjsjann123 commented:
> I think changes to `test_nvfuser_capability_context` should also be reverted in addition to `torch/_prims/context.py`.
> 
> Can you please provide a concrete example of what situation changes to `context.py` prevent? Because I don't see what you would like to achieve.

Updated per offline discussion. moved aten fallback logic to runtime execution instead of tracing. Patched tests as well.

jjsjann123 committed:
lintrunner

kevinstephano reviewed (approved):
Look good to me.

IvanYashchuk reviewed (approved):
Thank you for the changes!

IvanYashchuk commented:
@pytorchbot merge -g
